Source/WebCore/platform/graphics/BitmapImageDescriptor.h

```diff
@@ -0,0 +1,118 @@
+    IntSize sourceSize(ImageOrientation = ImageOrientation::Orientation::FromImage) const;
+    std::optional<IntSize> densityCorrectedSize() const;
+    ImageOrientation orientation() const;
+    unsigned primaryFrameIndex() const;
+    unsigned frameCount() const;
+    RepetitionCount repetitionCount() const;
+    DestinationColorSpace colorSpace() const;
+    std::optional<Color> singlePixelSolidColor() const;
+
+    String uti() const;
+    String filenameExtension() const;
+    String accessibilityDescription() const;
+    std::optional<IntPoint> hotSpot() const;
+    SubsamplingLevel maximumSubsamplingLevel() const;
+
+    SubsamplingLevel subsamplingLevelForScaleFactor(GraphicsContext&, const FloatSize& scaleFactor, AllowImageSubsampling) const;
+
+    void dump(TextStream&) const;
+
+private:
+    enum class CachedFlag : uint16_t {
+        EncodedDataStatus           = 1 << 0,
+        Size                        = 1 << 1,
+        DensityCorrectedSize        = 1 << 2,
+        Orientation                 = 1 << 3,
+        PrimaryFrameIndex           = 1 << 4,
+        FrameCount                  = 1 << 5,
+        RepetitionCount             = 1 << 6,
+        ColorSpace                  = 1 << 7,
+        SinglePixelSolidColor       = 1 << 8,
+
+        UTI                         = 1 << 9,
+        FilenameExtension           = 1 << 10,
+        AccessibilityDescription    = 1 << 11,
+        HotSpot                     = 1 << 12,
+        MaximumSubsamplingLevel     = 1 << 13,
+    };
+
+    template<typename MetadataType>
+    MetadataType imageMetadata(MetadataType& cachedValue, const MetadataType& defaultValue, CachedFlag, MetadataType (ImageDecoder::*functor)() const) const;
+
+    template<typename MetadataType>
+    MetadataType primaryImageFrameMetadata(MetadataType& cachedValue, CachedFlag, MetadataType (ImageFrame::*functor)() const) const;
+
+    template<typename MetadataType>
+    MetadataType primaryNativeImageMetadata(MetadataType& cachedValue, const MetadataType& defaultValue, CachedFlag, MetadataType (NativeImage::*functor)() const) const;
+
+    mutable OptionSet<CachedFlag> m_cachedFlags;
+
+    mutable EncodedDataStatus m_encodedDataStatus { EncodedDataStatus::Unknown };
+    mutable IntSize m_size;
+    mutable std::optional<IntSize> m_densityCorrectedSize;
+    mutable ImageOrientation m_orientation;
+    mutable size_t m_primaryFrameIndex { 0 };
+    mutable size_t m_frameCount { 0 };
+    mutable RepetitionCount m_repetitionCount { RepetitionCountNone };
+    mutable DestinationColorSpace m_colorSpace { DestinationColorSpace::SRGB() };
+    mutable std::optional<Color> m_singlePixelSolidColor;
+
+    mutable String m_uti;
+    mutable String m_filenameExtension;
+    mutable String m_accessibilityDescription;
+    mutable std::optional<IntPoint> m_hotSpot;
+    mutable SubsamplingLevel m_maximumSubsamplingLevel { SubsamplingLevel::Default };
+
+    BitmapImageSource& m_source;
```

Comment:
Per https://a1391192.slack.com/archives/GK0ATJ6FJ/p1707106610975729 can you make this a WeakRef?

Reply:
This caused a MM regression in the Leaves sub-test. So I had to take it out. Maybe after the recent changes in `BitmapImage` class and its friends, this is not true anymore.

So I will address this in a separate a separate patch after making sure it does not cause any regression.